### ISSUE TITLE
Update to ubuntu 18 seems to have broken the software (tensorflow version issue?).

### ISSUE THREAD

**Issue by btemperton** [ashsmith88/IMPACT2AMR_galleria_imaging#8]
Currently, the software throws no errors, but it is failing to find the boundaries of the Galleria (see image). I think part of the problem is that the previous version of tensorflow and associated CUDA libraries are not supported on ubuntu 18 and so have been updated. I wonder if the `Galleria.h5` model was created with a version that is not compatible with current versions. The commands I used to reinstall the software were:

```
mamba create -n galleria -c conda-forge python=3.7.2
conda activate galleria
git clone https://github.com/ashsmith88/IMPACT2AMR_galleria_imaging.git
cd IMPACT2AMR_galleria_imaging/
pip install gdown
gdown https://drive.google.com/uc?id=1dK_C3ExiqM4uABvMa59CH7-gfNG8K0OG
pip install --upgrade pip setuptools
pip install -r requirements.txt
pip install bigtiff
```
Then I ran:
```
python -m analysis_suite test-imgs/ -plate rect50 -batch
```
![bioluminescence_reading_T0](https://user-images.githubusercontent.com/1521603/151158782-f85e999f-08f0-4775-b4f6-b5b0f6848703.jpg)

Running the following in python shows `tensorflow` can find the GPUs
```
Python 3.7.2 | packaged by conda-forge | (default, Mar 20 2019, 01:46:52) 
[GCC 7.3.0] :: Anaconda, Inc. on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import tensorflow as tf
>>> tf.config.list_physical_devices('GPU')
[PhysicalDevice(name='/physical_device:GPU:0', device_type='GPU'), PhysicalDevice(name='/physical_device:GPU:1', device_type='GPU'), PhysicalDevice(name='/physical_device:GPU:2', device_type='GPU'), PhysicalDevice(name='/physical_device:GPU:3', device_type='GPU')]
```

`nvidia-smi` gives the following:
```
nvidia-smi
Wed Jan 26 11:56:26 2022       
+-----------------------------------------------------------------------------+
| NVIDIA-SMI 465.19.01    Driver Version: 465.19.01    CUDA Version: 11.3     |
|-------------------------------+----------------------+----------------------+
| GPU  Name        Persistence-M| Bus-Id        Disp.A | Volatile Uncorr. ECC |
| Fan  Temp  Perf  Pwr:Usage/Cap|         Memory-Usage | GPU-Util  Compute M. |
|                               |                      |               MIG M. |
|===============================+======================+======================|
|   0  NVIDIA GeForce ...  On   | 00000000:01:00.0  On |                  N/A |
|  0%   26C    P8    26W / 250W |    185MiB / 11011MiB |      0%      Default |
|                               |                      |                  N/A |
+-------------------------------+----------------------+----------------------+
|   1  NVIDIA GeForce ...  On   | 00000000:21:00.0 Off |                  N/A |
|  0%   27C    P8    40W / 250W |      1MiB / 11019MiB |      0%      Default |
|                               |                      |                  N/A |
+-------------------------------+----------------------+----------------------+
|   2  NVIDIA GeForce ...  On   | 00000000:4B:00.0 Off |                  N/A |
|  0%   28C    P8    20W / 250W |      1MiB / 11019MiB |      0%      Default |
|                               |                      |                  N/A |
+-------------------------------+----------------------+----------------------+
|   3  NVIDIA GeForce ...  On   | 00000000:4C:00.0 Off |                  N/A |
|  0%   26C    P8     8W / 250W |      1MiB / 11019MiB |      0%      Default |
|                               |                      |                  N/A |
+-------------------------------+----------------------+----------------------+

```

The `conda` environment being used is:
```
name: galleria
channels:
  - conda-forge
  - defaults
dependencies:
  - _libgcc_mutex=0.1=conda_forge
  - _openmp_mutex=4.5=1_gnu
  - bzip2=1.0.8=h7f98852_4
  - ca-certificates=2021.10.8=ha878542_0
  - libffi=3.2.1=he1b5a44_1007
  - libgcc-ng=11.2.0=h1d223b6_12
  - libgomp=11.2.0=h1d223b6_12
  - libstdcxx-ng=11.2.0=he4da1e4_12
  - libzlib=1.2.11=h36c2ea0_1013
  - ncurses=6.3=h9c3ff4c_0
  - openssl=1.1.1l=h7f98852_0
  - pip=21.3.1=pyhd8ed1ab_0
  - python=3.7.2=h381d211_0
  - python_abi=3.7=2_cp37m
  - readline=7.0=hf8c457e_1001
  - setuptools=60.5.0=py37h89c1867_0
  - sqlite=3.28.0=h8b20d00_0
  - tk=8.6.11=h27826a3_1
  - wheel=0.37.1=pyhd8ed1ab_0
  - xz=5.2.5=h516909a_1
  - zlib=1.2.11=h36c2ea0_1013
  - pip:
    - absl-py==1.0.0
    - astunparse==1.6.3
    - beautifulsoup4==4.10.0
    - bigtiff==0.1.2
    - cached-property==1.5.2
    - cachetools==5.0.0
    - certifi==2021.10.8
    - charset-normalizer==2.0.10
    - cycler==0.11.0
    - docopt==0.6.2
    - filelock==3.4.2
    - flatbuffers==2.0
    - gast==0.4.0
    - gdown==4.2.0
    - google-auth==2.5.0
    - google-auth-oauthlib==0.4.6
    - google-pasta==0.2.0
    - grpcio==1.43.0
    - h5py==3.6.0
    - idna==3.3
    - imageio==2.14.1
    - importlib-metadata==4.10.1
    - joblib==1.1.0
    - kaitaistruct==0.8
    - keras==2.7.0
    - keras-preprocessing==1.1.2
    - kiwisolver==1.3.2
    - libclang==12.0.0
    - llvmlite==0.38.0
    - markdown==3.3.6
    - matplotlib==3.4.3
    - munkres==1.1.4
    - networkx==2.6.3
    - numba==0.55.0
    - numpy==1.21.5
    - oauthlib==3.1.1
    - opt-einsum==3.3.0
    - pandas==1.3.2
    - pillow==9.0.0
    - protobuf==3.19.3
    - pyasn1==0.4.8
    - pyasn1-modules==0.2.8
    - pyparsing==3.0.7
    - pyqt5==5.15.4
    - pyqt5-qt5==5.15.2
    - pyqt5-sip==12.9.0
    - pysocks==1.7.1
    - python-dateutil==2.8.2
    - pytz==2021.3
    - pywavelets==1.2.0
    - requests==2.27.1
    - requests-oauthlib==1.3.0
    - rsa==4.8
    - scikit-image==0.18.3
    - scikit-learn==1.0.2
    - scipy==1.7.3
    - six==1.16.0
    - sklearn==0.0
    - soupsieve==2.3.1
    - sparse==0.13.0
    - tensorboard==2.8.0
    - tensorboard-data-server==0.6.1
    - tensorboard-plugin-wit==1.8.1
    - tensorflow==2.7.0
    - tensorflow-estimator==2.7.0
    - tensorflow-gpu==2.7.0
    - tensorflow-io-gcs-filesystem==0.23.1
    - termcolor==1.1.0
    - threadpoolctl==3.0.0
    - tifffile==2021.11.2
    - tqdm==4.62.3
    - typing-extensions==4.0.1
    - urllib3==1.26.8
    - werkzeug==2.0.2
    - wrapt==1.13.3
    - yapic==1.0.1
    - yapic-io==0.2.7
    - zipp==3.7.0
```

The version of `cuDNN` is `8.3.2.44_1.`

**Comment by ashsmith88** [comment]
@btemperton 
I haven't worked on this for quite a while now, but I am pretty sure that I set it up so that the model should still predict on CPU as well as GPU, therefore, I am more inclined to think that the issue is actually the model is not being correctly loaded into the program at all. 

There is a try / except step and I think this is likely where it is failing to load. Off the top of my head I can't remember exactly which folder the model file should be in (I believe its the project root), but perhaps try it in both the project root folder and also try it with the model saved in the analysis_suite folder? Also, make sure it is named properly: `Galleria_model.h5` 

If the issue persists, then I would look at `run_model` in `run_yapic_model.py` - perhaps debug the try / except statement to determine where it is failing on the predicition.

**Comment by btemperton** [comment]
Thanks Ashley, hardcoding the location of `Galleria_model.h5` in `run_yapic_model.py` doesn't make any difference. However, when I add the following exception code to `run_model` in `run_yapic_model.py`:

```
def run_model(images):
    buff = 0
    while buff <= 1:
        try:
            tf.keras.backend.clear_session()
            converted_images, zoom_factor = convert_image_size(images, buffer=buff)
            model = "/home/robyn/galleria/IMPACT2AMR_galleria_imaging/Galleria_model.h5"
            session = Session()
            session.dataset = load_connector(converted_images)
            session.load_model(model)
            print(f'Loaded model: {model}')
            session.set_normalization('local')
            all_wells = predict(session)
            all_wells.shrink_images(zoom_factor)
            return all_wells
            break
        except ValueError as ve:
            print('Handling Value error:', ve)
        except Exception as e:
            print("Unexpected error:", sys.exc_info()[0])   
            buff += 0.25
            continue
    return None
```
I get the following message:
```
Handling Value error: Memory growth cannot differ between GPU devices
```

This is being thrown by
```
session.load_model(model)
```
as a print statement under that line is never reached.

Any ideas how to fix that?

**Comment by btemperton** [comment]
Cool, that fixed that problem! Thanks! Now
```
session.load_model(model)
```
is throwing a Type Error stating that the `Session` object from `yapic` requires a `batch_size` positional argument
```
Handling TypeError: __init__() missing 1 required positional argument: 'batch_size'
Traceback (most recent call last):
  File "/home/robyn/galleria/IMPACT2AMR_galleria_imaging/analysis_suite/run_yapic_model.py", line 33, in run_model
    session.load_model(model)
  File "/home/robyn/miniconda3/envs/galleria/lib/python3.7/site-packages/yapic/session.py", line 133, in load_model
    output_tile_size_zxy)
  File "/home/robyn/miniconda3/envs/galleria/lib/python3.7/site-packages/yapic/session.py", line 145, in _configure_minibatch_data
    padding_zxy=padding_zxy)
```

This is curious because the `__init__` method of `yapic.Session` `v1.01` doesn't contain `batch_size`, but the latest version does. However, when I run `pip list` in my environment to see which version it is using, it says

```
yapic                        1.0.1
yapic-io                     0.2.7
```

**Comment by ashsmith88** [comment]
@btemperton 
The issue isn't in the __init__() of session, its in the init of `TrainingBatch` which is imported from a separate module called yapic_io. Previous versions of this didn't have `batch_size` in the`__init__()` but the most recent release (about 4 weeks ago) now has. 

The problem appears to be that we don't specify the version of yapic_io in requirements.txt. 
Try the following

- Uninstall yapic_io from your venv
- Add `yapic_io==0.2.6` to requirements.txt (This is the earliest version and timescales roughly aline with our yapic version)
- Run `pip install -r requirements.txt` to update your venv and install the old version of yapic_io

Hope that helps. If it does, this is likely the route cause of your problem (and not the GPU issue)

**Comment by btemperton** [comment]
Success! Thanks Ashley :). Simply removing `yapic-io` using `pip uninstall` and then `pip install yapic_io==0.2.6` resolved the issue.
![bioluminescence_reading_T0](https://user-images.githubusercontent.com/1521603/151524344-95b0317b-e7bb-4729-b52e-d8da460a97e0.jpg)

I've attached my updated `run_yapic_model.py` script and `requirements.txt` for you to use as you see fit. The script now handles multiple GPUs, various error types and implements proper logging. The `requirements.txt` includes the proper version of `yapic-io` and `bigtiff`, which was also needed and missing after install.

[update.tar.gz](https://github.com/ashsmith88/IMPACT2AMR_galleria_imaging/files/7957561/update.tar.gz)

**Comment by ashsmith88** [comment]
Excellent. No problem. Will add some changes to the code base, thanks for raising the issue. 

Closed by: b1922fbcdc0a1eab391502346a7c2745cfe989f4
